src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -63,4 +63,96 @@ public void testFlushOnDeleteSetting() throws Exception {
+
+            ShardStateMetaData state2 = new ShardStateMetaData(version, primary, "foo");
+            ShardStateMetaData.write(logger, "foo", id, state2, randomBoolean(), env.shardPaths(id));
+            shardStateMetaData = ShardStateMetaData.load(logger, id, env.shardPaths(id));
+            assertEquals(shardStateMetaData, state1);
+
+            ShardStateMetaData state3 = new ShardStateMetaData(version + 1, primary, "foo");
+            ShardStateMetaData.write(logger, "foo", id, state3, randomBoolean(), env.shardPaths(id));
+            shardStateMetaData = ShardStateMetaData.load(logger, id, env.shardPaths(id));
+            assertEquals(shardStateMetaData, state3);
+            assertEquals("foo", state3.indexUUID);
+        }
+    }
+
+    public void testPersistenceStateMetadataPersistence() throws Exception {
+        createIndex("test");
+        ensureGreen();
+        IndicesService indicesService = getInstanceFromNode(IndicesService.class);
+        NodeEnvironment env = getInstanceFromNode(NodeEnvironment.class);
+        IndexService test = indicesService.indexService("test");
+        IndexShard shard = test.shard(0);
+        ShardStateMetaData shardStateMetaData = ShardStateMetaData.load(logger, shard.shardId, env.shardPaths(shard.shardId));
+        assertEquals(getShardStateMetadata(shard), shardStateMetaData);
+        ShardRouting originalRouting = shard.shardRouting;
+        ShardRouting routing = new MutableShardRouting(shard.shardRouting, shard.shardRouting.version()+1);
+        shard.updateRoutingEntry(routing, true);
+
+        shardStateMetaData = ShardStateMetaData.load(logger, shard.shardId, env.shardPaths(shard.shardId));
+        assertEquals(shardStateMetaData, getShardStateMetadata(shard));
+
+        routing = new MutableShardRouting(shard.shardRouting, shard.shardRouting.version()+1);
+        shard.updateRoutingEntry(routing, true);
+        shardStateMetaData = ShardStateMetaData.load(logger, shard.shardId, env.shardPaths(shard.shardId));
+        assertEquals(shardStateMetaData, getShardStateMetadata(shard));
+
+        shard.updateRoutingEntry(new MutableShardRouting(originalRouting, shard.shardRouting.version() + 1), true);
+        shardStateMetaData = ShardStateMetaData.load(logger, shard.shardId, env.shardPaths(shard.shardId));
+        assertEquals(shardStateMetaData, getShardStateMetadata(shard));
+        // test if we still write it even if the shard is not active
+        MutableShardRouting inactiveRouting = new MutableShardRouting(shard.shardRouting.index(), shard.shardRouting.shardId().id(), shard.shardRouting.currentNodeId(), true, ShardRoutingState.INITIALIZING, shard.shardRouting.version() + 1);
+        shard.persistMetadata(inactiveRouting, getShardStateMetadata(shard));
+        shardStateMetaData = ShardStateMetaData.load(logger, shard.shardId, env.shardPaths(shard.shardId));
+        assertEquals(shardStateMetaData, getShardStateMetadata(shard));
```

Comment:
can we add a failure message saying "inactive shard state shouldn't be persisted".